File: 2020-03-13-haircut/hunch-confirmer.cc
#include <deque>
#include <iostream>
#include <random>

using Generator = std::minstd_rand;

bool next_in_line;
long waiting = 0;

std::bernoulli_distribution dist(0.25);

void push_new(Generator& rng) {
    next_in_line = dist(rng);
}

void forward(const int barber, Generator& rng) {
    if (barber == 0) {  // It's Tiffany
        if (waiting > 0)
            waiting--;
        else {
            push_new(rng);
        }
    } else {
        while (next_in_line == true) {
            waiting++;
            push_new(rng);
        }

        push_new(rng);
    }
}

int main() {
    Generator rng{std::random_device{}()};

    push_new(rng);

    long total = 0;
    long non_tiffanys = 0;

    while (true) {
        for (int i = 0; i < 4; i++) {
            total++;
            if (waiting == 0 && next_in_line != true) {
                non_tiffanys++;
            }

            if (total % long(1e8) == 0) {
                std::cout << non_tiffanys << " " << total << " "
                          << std::round(std::sqrt(total)) << " "
                          << non_tiffanys / double(total) << std::endl;
            }

            forward(i, rng);
        }
    }
}

File: 2020-03-13-haircut/hunch-confirmer.py
#!/usr/bin/env python3

import math
import random
import itertools
import numpy as np
import enum
from collections import deque

class Barbers(enum.Enum):
    TIFFANY = 0
    B1 = 1
    B2 = 2
    B3 = 3

def forward(line, waiting_for_tiffany, barber):
    if barber == Barbers.TIFFANY:
        if waiting_for_tiffany > 0:
            waiting_for_tiffany -= 1
        else:
            line.popleft()
    else:
        while line and line[0] == Barbers.TIFFANY:
            waiting_for_tiffany += 1
            line.popleft()
        if line:
            line.popleft()
    return line, waiting_for_tiffany


def push_new(line, waiting_for_tiffany):
    line.append(random.choice(list(Barbers)))
    return line, waiting_for_tiffany

line = deque()
waiting_for_tiffany = 0

for i in range(10):
    line, waiting_for_tiffany = push_new(line, waiting_for_tiffany)

total = 0
non_tiffanys = 0
for b in itertools.cycle(random.sample(list(Barbers), len(Barbers))):
    total += 1
    if waiting_for_tiffany == 0 and line[0] != Barbers.TIFFANY:
        non_tiffanys += 1

    if total % 100000 == 0:
        print(non_tiffanys, total, int(math.sqrt(total)), non_tiffanys / total)
    # print(b, [line[i].name for i in range(5)], waiting_for_tiffany)
    line, waiting_for_tiffany = forward(line, waiting_for_tiffany, b)
    while len(line) < 10:
        line, waiting_for_tiffany = push_new(line, waiting_for_tiffany)
